# **Features Inside the Courtroom**

**Sources:**

Supreme Court Oral Arguments Corpus: https://convokit.cornell.edu/documentation/supreme.html

In [1]:
#install convokit
!pip install convokit

#import and download corpus
from convokit import Corpus, download
corpus = Corpus(filename=download("supreme-corpus"))

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 240.9/240.9 kB 4.0 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.4/5.4 MB 55.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 331.1/331.1 kB 11.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 31.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.8/235.8 kB 9.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 608.4/608.4 kB 19.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.8/44.8 kB 1.3 MB/s eta 0:00:00
  Created wheel for convokit: filename=convokit-4.1.0-py3-none-any.whl size=292185 sha256=61b79d3f83fa8baf8e7f158c967757e0dde522b4ccbb52acdca67a75ef360b29
  Stored in directory: /root/.cache/pip/wheels/5e/d8/32/b867fe29313972bfba4cb

In [2]:
import pandas as pd
cases_df = pd.read_json("https://zissou.infosci.cornell.edu/convokit/datasets/supreme-corpus/cases.jsonl", lines=True)
cases_df.head()


,id,year,citation,title,petitioner,respondent,docket_no,court,decided_date,url,...,adv_sides_inferred,known_respondent_adv,advocates,win_side,win_side_detail,scdb_docket_id,votes,votes_detail,is_eq_divided,votes_side
0,1955_71,1955,350 US 79,Affronti v. United States,Affronti,United States,71,Warren Court,"Dec 5, 1955",https://www.oyez.org/cases/1955/71,...,True,True,"{'Harry F. Murphy': {'id': 'harry_f_murphy', '...",0.0,2.0,1955-009-01,"{'j__john_m_harlan2': 2.0, 'j__hugo_l_black': ...","{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ...",0.0,"{'j__john_m_harlan2': 0.0, 'j__hugo_l_black': ..."
1,1955_410,1955,351 US 79,"American Airlines, Inc. v. North American Airl...","American Airlines, Inc.","North American Airlines, Inc.",410,Warren Court,"Apr 23, 1956",https://www.oyez.org/cases/1955/410,...,True,True,{'Howard C. Westwood': {'id': 'howard_c_westwo...,1.0,4.0,1955-071-01,"{'j__john_m_harlan2': 2.0, 'j__hugo_l_black': ...","{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ...",0.0,"{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ..."
2,1955_351,1955,350 US 532,Archawski v. Hanioti,Archawski,Hanioti,351,Warren Court,"Apr 9, 1956",https://www.oyez.org/cases/1955/351,...,True,False,"{'Harry D. Graham': {'id': 'harry_d_graham', '...",1.0,4.0,1955-053-01,"{'j__john_m_harlan2': 2.0, 'j__hugo_l_black': ...","{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ...",0.0,"{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ..."
3,1955_38,1955,350 US 568,Armstrong v. Armstrong,Armstrong,Armstrong,38,Warren Court,"Apr 9, 1956",https://www.oyez.org/cases/1955/38,...,True,False,"{'Robert N. Gorman': {'id': 'robert_n_gorman',...",0.0,2.0,1955-056-01,"{'j__john_m_harlan2': 2.0, 'j__hugo_l_black': ...","{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ...",0.0,"{'j__john_m_harlan2': 0.0, 'j__hugo_l_black': ..."
4,1955_49,1955,350 US 198,"Bernhardt v. Polygraphic Company of America, Inc.",Bernhardt,"Polygraphic Company of America, Inc.",49,Warren Court,"Jan 16, 1956",https://www.oyez.org/cases/1955/49,...,True,False,"{'Manfred W. Ehrich, Jr.': {'id': 'manfred_w_e...",1.0,4.0,1955-020-01,"{'j__john_m_harlan2': 1.0, 'j__hugo_l_black': ...","{'j__john_m_harlan2': 2.0, 'j__hugo_l_black': ...",0.0,"{'j__john_m_harlan2': 0.0, 'j__hugo_l_black': ..."


**Separate components of corpus:**
- Utterances
- Converstions
- Speakers

In [3]:
#UTTERANCES
utterances_df = corpus.get_utterances_dataframe()
utterances_df.head()

,timestamp,text,speaker,reply_to,conversation_id,meta.case_id,meta.start_times,meta.stop_times,meta.speaker_type,meta.side,meta.timestamp,vectors
id,,,,,,,,,,,,
13127__0_000,None,"Number 71, Lonnie Affronti versus United State...",j__earl_warren,None,13127,1955_71,"[0.0, 7.624]","[7.624, 9.218]",J,None,0.0,[]
13127__0_001,None,May it please the Court.\nWe are here by writ ...,harry_f_murphy,13127__0_000,13127,1955_71,"[9.218, 11.538, 15.653, 22.722, 28.849, 33.575]","[11.538, 15.653, 22.722, 28.849, 33.575, 48.138]",A,1,9.218,[]
13127__0_002,None,Consecutive sentences.,j__william_o_douglas,13127__0_001,13127,1955_71,[48.138],[49.315],J,None,48.138,[]
13127__0_003,None,"Consecutive sentences.\nIn this case, the defe...",harry_f_murphy,13127__0_002,13127,1955_71,"[49.315, 51.844, 60.81, 67.083, 72.584, 89.839...","[51.844, 60.81, 67.083, 72.584, 89.839, 95.873...",A,1,49.315,[]
13127__0_004,None,Was the aggregate prison sentence was 20 or 25...,<INAUDIBLE>,13127__0_003,13127,1955_71,[174.058],[176.766],None,None,174.058,[]


In [4]:
#CONVERSATIONS
conversations_df = corpus.get_conversations_dataframe()
conversations_df.head()

,vectors,meta.case_id,meta.advocates,meta.win_side,meta.votes_side
id,,,,,
13127,[],1955_71,"{'harry_f_murphy': {'side': 1, 'role': 'inferr...",0,"{'j__john_m_harlan2': 0, 'j__hugo_l_black': 0,..."
12997,[],1955_410,"{'howard_c_westwood': {'side': 1, 'role': 'inf...",1,"{'j__john_m_harlan2': 1, 'j__hugo_l_black': 1,..."
13024,[],1955_410,"{'howard_c_westwood': {'side': 1, 'role': 'inf...",1,"{'j__john_m_harlan2': 1, 'j__hugo_l_black': 1,..."
13015,[],1955_351,"{'harry_d_graham': {'side': 3, 'role': 'inferr...",1,"{'j__john_m_harlan2': 1, 'j__hugo_l_black': 1,..."
13016,[],1955_38,"{'robert_n_gorman': {'side': 3, 'role': 'infer...",0,"{'j__john_m_harlan2': 0, 'j__hugo_l_black': 0,..."


In [5]:
#SPEAKERS
speakers_df = corpus.get_speakers_dataframe()
speakers_df.head()

,vectors,meta.name,meta.type,meta.role
id,,,,
j__earl_warren,[],Earl Warren,J,justice
harry_f_murphy,[],Harry F. Murphy,A,NaN
j__william_o_douglas,[],William O. Douglas,J,justice
<INAUDIBLE>,[],<INAUDIBLE>,U,inaudible
j__felix_frankfurter,[],Felix Frankfurter,J,justice


**Analysis:**
Winning vs Losing side features:
- Sentiment Score
- Lawyer Response Length to Justice's Questions
- Citing precedent

In [6]:
#winning side distribution
print(utterances_df["meta.side"].value_counts())
print(utterances_df["meta.speaker_type"].value_counts())

meta.side
1    341431
0    305071
3    154335
2     24432
Name: count, dtype: int64
meta.speaker_type
A    825269
J    787417
Name: count, dtype: int64


In [7]:
conversations_df["meta.win_side"].value_counts()

,count
meta.win_side,
1,4930
0,2868
2,4
-1,2


*Interpretation*

---



1= how many cases were won by the petitioner (4930)

0= how many cases were won by the respondent (2868)

-1= how many cases have no winner information (2)


In [8]:
#get all the advocate utterances and save them as their own separate dataframe
advocate_utterances = utterances_df[utterances_df["meta.speaker_type"] == "A"].copy()

#look up winning side for each utterance based on which conversation it came from
advocate_utterances = advocate_utterances.merge(
    conversations_df[["meta.win_side"]],
    left_on="conversation_id",
    right_index=True,
    how="left"
)

advocate_utterances.head()

,timestamp,text,speaker,reply_to,conversation_id,meta.case_id,meta.start_times,meta.stop_times,meta.speaker_type,meta.side,meta.timestamp,vectors,meta.win_side
id,,,,,,,,,,,,,
13127__0_001,None,May it please the Court.\nWe are here by writ ...,harry_f_murphy,13127__0_000,13127,1955_71,"[9.218, 11.538, 15.653, 22.722, 28.849, 33.575]","[11.538, 15.653, 22.722, 28.849, 33.575, 48.138]",A,1,9.218,[],0
13127__0_003,None,"Consecutive sentences.\nIn this case, the defe...",harry_f_murphy,13127__0_002,13127,1955_71,"[49.315, 51.844, 60.81, 67.083, 72.584, 89.839...","[51.844, 60.81, 67.083, 72.584, 89.839, 95.873...",A,1,49.315,[],0
13127__0_005,None,Twenty years.,harry_f_murphy,13127__0_004,13127,1955_71,[176.766],[177.525],A,1,176.766,[],0
13127__0_007,None,"Twenty years, four -- four counts of five year...",harry_f_murphy,13127__0_006,13127,1955_71,"[177.844, 183.613, 204.06, 225.112, 253.196, 2...","[183.613, 204.06, 225.112, 253.196, 268.426, 2...",A,1,177.844,[],0
13127__0_009,None,"Well the -- the probation statute, I think, wh...",harry_f_murphy,13127__0_008,13127,1955_71,[566.261],[576.08],A,1,566.261,[],0


*Interpretation*


---


advocate_utterances:
has all the advocate speech, but doesn't know who won


conversations_df[["meta.win_side"]]:
has the winning side information for each conversation


The merge connects them by matching conversation_id in advocate_utterances (every utterance knows which conversation it belongs to) with the index of conversations_df (which is the conversation ID). So after the merge, every advocate utterance will have a new column meta.win_side telling us who won that case. Left merge keeps all the advocate utterances even if no match is found.

In [9]:
#create won column
advocate_utterances["won"] = (
    advocate_utterances["meta.side"] == advocate_utterances["meta.win_side"]
).astype(int)

#drop -1 cases
advocate_utterances = advocate_utterances[advocate_utterances["meta.win_side"] != -1]

advocate_utterances["won"].value_counts()

,count
won,
0,514974
1,310202


*Interpretation*


---



advocate_utterances["meta.side"]:
tells which side the advocate is on (1 for petitioner, 0 for respondent)

advocate_utterances["meta.win_side"]:
tells us which side won the case (1 for petitioner, 0 for respondent)


- So if an advocate is on side 1 and the winning side is 1, they match and the advocate won.

- If an advocate is on side 0 and the winning side is 1, they don't match and the advocate lost.

- The comparison gives us 1's or 0's indicating whether an advocate was on the winning side (1) or on the losing side (0).


In [17]:
#average sentiment scores for winning vs losing advocates
from nltk.sentiment import SentimentIntensityAnalyzer
import nltk
nltk.download('vader_lexicon')

sia = SentimentIntensityAnalyzer()

# get sentiment score for each utterance
sentiments = []
for text in advocate_utterances["text"]:
    #compound = VADER's overall sentiment score for a piece of text, a single number between -1.0 and +1.0
    score = sia.polarity_scores(str(text))["compound"]
    sentiments.append(score)

advocate_utterances.loc[:, "sentiment"] = sentiments

advocate_utterances.groupby("won")["sentiment"].mean()

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


,sentiment
won,
0,0.175560
1,0.178376


**Findings:**

Sentiment scores are nearly identical between winning and losing advocates which suggests that positivity of language is not a distinguishing factor in Supreme Court oral arguments. Therefore, word choice matters more than tone.

In [21]:
import plotly.express as px

sentiment_means = advocate_utterances.groupby("won")["sentiment"].mean().reset_index()
sentiment_means["outcome"] = ["Losing", "Winning"]

fig = px.bar(sentiment_means, x="outcome", y="sentiment",
             color="outcome",
             color_discrete_map={"Losing": "lightskyblue", "Winning": "#5B2D8E"},
             title="Average Sentiment Score by Outcome",
             labels={"sentiment": "Average Sentiment Score", "outcome": ""})

fig.update_layout(showlegend=False)

#update bar properties to display the sentiment to the 4th decimal
fig.update_traces(texttemplate="%{y:.4f}", textposition="outside")
fig.show()

In [23]:
advocate_utterances["year"] = advocate_utterances["meta.case_id"].str.split("_").str[0].astype(int)
advocate_utterances["decade"] = (advocate_utterances["year"] // 10) * 10
advocate_utterances["won_label"] = advocate_utterances["won"].map({0: "Losing", 1: "Winning"})

sentiment_by_decade = advocate_utterances.groupby(["decade", "won_label"])["sentiment"].mean().reset_index()

fig = px.line(sentiment_by_decade, x="decade", y="sentiment",
              color="won_label",
              markers=True,
              color_discrete_map={"Losing": "lightskyblue", "Winning": "#5B2D8E"},
              title="Average Sentiment Score by Decade and Outcome",
              labels={"sentiment": "Average Sentiment Score", "decade": "Decade", "won_label": ""})

fig.update_layout(width=1500)
fig.show()

In [11]:
#winning advocate response length
response_lengths = []
for text in advocate_utterances["text"]:
    #split on white space
    length = len(str(text).split())
    response_lengths.append(length)

#get all rows and one column
advocate_utterances.loc[:, "response_length"] = response_lengths

#get justice utterance IDs
justice_ids = set(utterances_df[utterances_df["meta.speaker_type"] == "J"].index)

#filter to advocate utterances that are replying to a justice
responses_to_justices = advocate_utterances[
    advocate_utterances["reply_to"].isin(justice_ids)
]

responses_to_justices.groupby("won")["response_length"].mean()



,response_length
won,
0,60.791526
1,65.736386


**Findings:**

Winning advocates tend to give longer responses to justice questions on average (65 words) compared to losing advocates (60 words).

In [24]:
response_means = advocate_utterances.groupby("won")["response_length"].mean().reset_index()
response_means["outcome"] = ["Losing", "Winning"]

fig = px.bar(response_means, x="response_length", y="outcome",
             orientation="h",
             color="outcome",
             color_discrete_map={"Losing": "lightskyblue", "Winning": "#5B2D8E"},
             title="Average Response Length by Outcome",
             labels={"response_length": "Average Words per Question", "outcome": ""})

fig.update_traces(texttemplate="%{x:.1f}", textposition="outside")
fig.update_layout(showlegend=False, width=1500)
fig.show()

In [14]:
#citing precedent average
citation_counts = []

for text in advocate_utterances["text"]:
    text = str(text).lower()
    count = text.count(" v. ")
    citation_counts.append(count)

advocate_utterances["citation_count"] = citation_counts
advocate_utterances["citation_rate"] = advocate_utterances["citation_count"] / advocate_utterances["response_length"]

advocate_utterances.groupby("won")["citation_rate"].mean()

,citation_rate
won,
0,0.000240
1,0.000298


In [15]:
from scipy import stats

winning = advocate_utterances[advocate_utterances["won"] == 1]["citation_rate"]
losing = advocate_utterances[advocate_utterances["won"] == 0]["citation_rate"]

stats.ttest_ind(winning, losing)

TtestResult(statistic=np.float64(6.385889623946161), pvalue=np.float64(1.704939241771275e-10), df=np.float64(825174.0))

**Findings:**

The results show winning advocates cite slightly more per word (0.000298) than losing ones (0.000240), though the difference is very small. After doing a t-test to verify that the difference is indeed statistically significant, the difference is too small to conclude much. Lawyers on the winning side do use a bit more precedent, but not much.

In [25]:
citation_means = advocate_utterances.groupby("won")["citation_rate"].mean().reset_index()
citation_means["outcome"] = ["Losing", "Winning"]

fig = px.bar(citation_means, x="outcome", y="citation_rate",
             color="outcome",
             color_discrete_map={"Losing": "lightskyblue", "Winning": "#5B2D8E"},
             title="Average Citation Rate by Outcome",
             labels={"citation_rate": "Average Citation Rate", "outcome": ""})

fig.update_traces(texttemplate="%{y:.6f}", textposition="outside")
fig.update_layout(showlegend=False)
fig.show()

In [26]:
#create new dataframe with average citation rate for each decade and outcome combination
citation_by_decade = advocate_utterances.groupby(["decade", "won_label"])["citation_rate"].mean().reset_index()
fig = px.line(citation_by_decade, x="decade",
              y="citation_rate",
              color="won_label",
              markers=True,
              color_discrete_map={"Losing": "lightskyblue", "Winning": "#5B2D8E"},
              title="Average Citation Rate by Decade and Outcome",
              labels={"citation_rate": "Average Citation Rate", "decade": "Decade", "won_label": ""})

fig.update_layout(width=1500)
fig.show()

**Analysis:**

Logistic Regression Model to predict whether a lawyer will win or lose based on stretegic features

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

#make one row per advocate per case
features = advocate_utterances.groupby(["speaker", "meta.case_id", "won"]).agg(
    sentiment=("sentiment", "mean"),
    response_length=("response_length", "mean"),
    citation_rate=("citation_rate", "mean"),
).reset_index()

features = features.dropna(subset=["sentiment", "response_length", "citation_rate"])

X_all = features[["sentiment", "response_length", "citation_rate"]]
y_all = features["won"]

scores = cross_val_score(LogisticRegression(class_weight="balanced"),
                         X_all,
                         y_all,
                         cv=5,
                         scoring="accuracy")

#mean accuracy of all features combined
print("Mean accuracy: " + str(scores.mean().round(3)))


Mean accuracy: 0.514


*Interpretation:*


---

The model correctly predicted whether an advocate won or lost 51.4% of the time. This is only slightly better than randomly guessing (50%), so these features are not the strongest predicters of win rate.

In [28]:
features_list = ["sentiment", "response_length", "citation_rate"]

#loop through each feature
for feature in features_list:
    X_one = features[[feature]]
    y_one = features["won"]

    scores = cross_val_score(LogisticRegression(class_weight="balanced"),
                             X_one,
                             y_one,
                             cv=5,
                             scoring="accuracy")

    print(feature + ":")
    print("  Mean accuracy: " + str(scores.mean().round(3)))
    print("  Std: " + str(scores.std().round(3)))
    print()

sentiment:
  Mean accuracy: 0.497
  Std: 0.015

response_length:
  Mean accuracy: 0.512
  Std: 0.07

citation_rate:
  Mean accuracy: 0.613
  Std: 0.0



*Interpretation:*


---

- Sentiment had a 49.7% accuracy rate which is literally worse than random guessing, so there is no predictive power at all.

- Response length had a 51.2% accuracy which is barely above randomly guessing, also suggesting a very weak signal. The high standard deviation on response length (0.07) is also worth noting, because it is very inconsistent across folds, meaning it performs well on some splits and poorly on others, which suggests it's not a reliable predictor.

- Citation rate, on the other hand, had a 61.3% accuracy, which is the highest of the three, meaning that it is the stongest predictor of win rate.


Conclusion:
Advocates who cite more case law are more likely to win.


In [29]:
import plotly.express as px

df = pd.DataFrame({
    "feature": ["Sentiment", "Response Length", "Citation Rate"],
    "accuracy": [0.497, 0.512, 0.613]
})

fig = px.bar(df,
             x="accuracy",
             y="feature",
             color="feature",
             color_discrete_map={
                 "Sentiment": "#C9B8E8",
                 "Response Length": "#9B6FD4",
                 "Citation Rate": "#5B2D8E"
             },
             title="Predictive Accuracy by Feature",
             labels={"accuracy": "Mean Accuracy", "feature": ""})

fig.add_vline(x=0.5, line_dash="dash", annotation_text="Random baseline")
fig.update_layout(showlegend=False, height= 700, width= 1300, xaxis=dict(range=[0, 0.65]))
fig.show()

In [30]:
X_two = features[["response_length", "citation_rate"]]

scores = cross_val_score(LogisticRegression(class_weight="balanced"),
                         X_two,
                         y_all,
                         cv=5,
                         scoring="accuracy")

print("Mean accuracy: " + str(scores.mean().round(3)))
print("Std: " + str(scores.std().round(3)))

Mean accuracy: 0.512
Std: 0.07


**Findings:**

Citation rate is the strongest signal, and combining it with response length only introduces noise